Makefile

```diff
@@ -34,13 +55,138 @@ up: ## Runs the local containers (n=service name)
 	@docker-compose up -d $(n)
 
 down: ## Stops the local containers and removes them
-	@echo "$(P) Stopping client and server..."
+	@echo "$(P) Stopping and removing client and server..."
 	@docker-compose down
 
+stop: ## Stops the local containers
+	@echo "$(P) Stopping client and server..."
+	@docker-compose stop ${n}
+
 build: ## Builds the local containers (n=service name)
 	@echo "$(P) Building images..."
 	@docker-compose build --no-cache $(n)
 
 rebuild: ## Build the local contains (n=service name) and then start them after building
 	@make build n=$(n)
 	@make up n=$(n)
+
+##############################################################################
+# Development environment specefic commands
+##############################################################################
+
+restart-dev: ## Restart local docker environment (n=service name)
+	$(info Restart local docker environment)
+	@make stop-dev n=$(n)
+	@make up-dev n=$(n)
+
+refresh-dev: ## Recreates local docker environment (n=service name)
+	$(info Recreates local docker environment)
+	@make stop-dev n=$(n)
+	@make build-dev n=$(n)
+	@make up-dev n=$(n)
+
+up-dev: ## Runs the local containers (n=service name)
+	@echo "$(P) Running client and server..."
+	@docker-compose -f docker-compose.dev.yaml up -d $(n)
+
+down-dev: ## Stops the local containers and removes them
+	@echo "$(P) Stopping client and server..."
+	@docker-compose -f docker-compose.dev.yaml down
+
+stop-dev: ## Stops the local containers
+	@echo "$(P) Stopping client and server..."
+	@docker-compose -f docker-compose.dev.yaml stop ${n}
+
+build-dev: ## Builds the local containers (n=service name)
+	@echo "$(P) Building images..."
+	@docker-compose -f docker-compose.dev.yaml build --no-cache $(n)
+
+rebuild-dev: ## Build the local contains (n=service name) and then start them after building
+	@make build n=$(n)
+	@make up-dev n=$(n)
+
+exec: ## Access the development workspace (n=service name)
+	@echo "Shelling into local application..."
+	@docker exec -it $(n) bash
+
+logs: ## Access application logs (n=service name)
+	@echo "Watching logging output for local development container..."
+	@docker logs -f $(shell docker inspect --format="{{.Id}}" ${n})
+
+##############################################################################
+# Local development commands
+##############################################################################
+
+build-local-development:
+	@echo "+\n++ Building local development Docker image...\n+"
+	@make build-dev n=$(API_SERVICE)
+	@make build-dev n=$(FRONTEND_SERVICE)
+	
+build-local-production:
+	@echo "+\n++ Building local production Docker image...\n+"
+	@make build n=$(API_SERVICE)
+	@make build n=$(FRONTEND_SERVICE)
+
+run-local-development:
+	@echo "+\n++ Running development container locally\n+"
+	@make up-dev n=$(API_SERVICE)
+	@make up-dev n=$(FRONTEND_SERVICE)
+
+run-local-production:
```

Comment:
Is there a need to have this many make commands? Can we set it up so that local and production are as much in sync? At the root folder I can explicitly do "docker compose up api" if I only want to bring up the API container. 
Can this suffice? 

My trouble is many of the syntax and execution is similar. If one makes a change in syntax, they would need to know where all to make the updates and it may causing breaking change. 
Keep it simple for us to update the Makefile in the future?

Reply:
Fair point. Let me revisit on what can be reduced. I think there are utilities at top which are being used at the end. Making a single `make` command to run the whole project was the ideology here.

Reply:
If you have few mins, let's connect over teams? @antsand